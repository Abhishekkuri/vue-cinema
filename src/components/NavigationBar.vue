<template>
  <div class="navigation-container">
    <nav>
      <router-link class="nav-link" to="/">Featured Films</router-link>
      <input
        type="text"
        v-model="title"
        placeholder="Search for a Movie"
        class="nav-input"
      />
      <button class="nav-button" @click="searchMovies">Search</button>
      <p class="error-message" v-if="searchedMovies.length === 0">
        Oops! try searching film with some other title.
      </p>
    </nav>
  </div>
</template>

<script>
import { mapGetters } from "vuex";
export default {
  data() {
    return {
      title: "",
    };
  },
  computed: {
    ...mapGetters({ searchedMovies: "getSearchedFilms" }),
  },
  methods: {
    async searchMovies() {
      this.showErrorMessage = false;
      await this.$store.dispatch("fetchSearchedFilms", {
        title: this.title,
      });
      this.$router.replace(`/search/${this.title}`);
    },
  },
};
</script>

<style lang="scss" scoped>
.navigation-container {
  background-color: $color-grey-light;
  nav {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: flex-start;
    font-size: 1.8rem;
    padding: 1rem;
    margin: 1rem;
    .nav-link {
      color: $color-black;
      border: 1px solid $color-white;
      text-decoration: none;
      padding: 1rem;
      margin-bottom: 4rem;
      border-radius: 10px;
      transition: 0.3s;
      &:hover {
        border: none;
        border: 1px solid $color-black;
        color: $color-grey-light;
        background-color: $color-primary;
        box-shadow: 0 8px 16px 0 $color-primary;
      }
    }
    .router-link-exact-active {
      color: $color-tirtiery;
      border-bottom: 1px solid $color-primary;
      border-radius: 0;
      border-top: none;
      border-left: 0;
      border-right: 0;
    }
    .nav-input {
      padding: 1rem;
      font-size: 1.5rem;
      border: 1px solid $color-tirtiery;
      border-radius: 10px;
      text-align: center;
      transition: 0.3s;
      &:hover {
        border: 1px solid $color-black;
        color: $color-grey-light;
        background-color: $color-primary;
        box-shadow: 0 8px 16px 0 $color-primary;
      }
    }
    .nav-button {
      padding: 1rem 2rem;
      font-size: 1.5rem;
      cursor: pointer;
      background-color: $color-grey-light;
      border: none;
      border-radius: 15px;
      margin-top: 1.5rem;
      transition: 0.3s;
      &:hover {
        border: 1px solid $color-black;
        color: $color-grey-light;
        background-color: $color-primary;
        box-shadow: 0 8px 16px 0 $color-primary;
      }
    }
  }
}

.error-message {
  color: red;
  font-size: 1rem;
  text-align: center;
}

@media screen and (max-width: 820px) {
  .navigation-container {
    nav {
      display: flex;
      flex-direction: row;
      align-items: center;
      justify-content: space-evenly;
      .nav-link {
        font-size: 1rem;
        padding: 1rem;
        margin: 0.5rem;
      }
      .nav-input {
        font-size: 1rem;
        padding: 1rem;
        margin: 0.5rem;
      }
      .nav-button {
        font-size: 1rem;
        padding: 0.5rem;
        margin: 0.5rem;
      }
    }
  }
}
</style>
