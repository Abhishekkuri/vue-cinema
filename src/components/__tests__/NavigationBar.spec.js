import { createLocalVue, RouterLinkStub, shallowMount } from "@vue/test-utils";
import Vuex from "vuex";
import NavigationBar from "@/components/NavigationBar";
import merge from "lodash.merge";
import flushPromises from "flush-promises";
const localVue = createLocalVue();

localVue.use(Vuex);

function createStore(overrides) {
  const defaultStoreConfig = {
    state: {
      featuredFilms: [{ film: {} }],
      films: [{ film: {} }],
    },
    getters: {
      getSearchedFilms: jest.fn(() => []),
    },
    actions: {
      fetchSearchedFilms: jest.fn(() => Promise.resolve()),
    },
  };
  return new Vuex.Store(merge(defaultStoreConfig, overrides));
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    stubs: {
      RouterLink: RouterLinkStub,
    },
    localVue,
    store: createStore(),
  };
  return shallowMount(NavigationBar, merge(defaultMountingOptions, overrides));
}

describe("NavigationBar.vue", () => {
  test("Navigation Bar renders correctly", () => {
    const wrapper = createWrapper();
    expect(wrapper.element).toMatchSnapshot();
  });
  test("Search button link works correctly", async () => {
    const mocks = {
      $router: {
        replace: jest.fn(),
      },
    };
    const store = createStore();
    store.dispatch = jest.fn(() => Promise.resolve());
    const title = "someTitle";
    const wrapper = createWrapper({ store, mocks });
    wrapper.setData({ title });
    wrapper.find(".nav-button").trigger("click");
    await flushPromises();
    expect(store.dispatch).toHaveBeenCalledWith("fetchSearchedFilms", {
      title,
    });
    expect(mocks.$router.replace).toHaveBeenCalledWith(`/search/${title}`);
  });
});
