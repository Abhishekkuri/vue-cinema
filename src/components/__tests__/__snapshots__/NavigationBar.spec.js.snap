// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationBar.vue Navigation Bar renders correctly 1`] = `
<div
  class="navigation-container"
>
  <nav>
    <a
      class="nav-link"
    >
      Featured Films
    </a>
     
    <input
      class="nav-input"
      placeholder="Search for a Movie"
      type="text"
    />
     
    <button
      class="nav-button"
    >
      Search
    </button>
  </nav>
</div>
`;
